include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,654 @@
+ * for our purposes. As the class or object goes out of scope, the communicator
+ * is freed.
+ */
+class communicator : public EnableSharedCreateMethod<communicator> {
+public:
+    communicator(const MPI_Comm& comm);
+
+    communicator(const MPI_Comm& comm, int color, int key);
+
+    communicator();
+
+    communicator(communicator& other);
+
+    communicator& operator=(const communicator& other);
+
+    communicator(communicator&& other);
+
+    communicator& operator=(communicator&& other);
+
+    static MPI_Comm get_comm_world() { return MPI_COMM_WORLD; }
+
+    static std::shared_ptr<communicator> create_world()
+    {
+        return std::make_shared<communicator>(get_comm_world());
+    }
+
+    MPI_Comm get() const { return comm_; }
+
+    int size() const { return size_; }
+
+    int rank() const { return rank_; };
+
+    int local_rank() const { return local_rank_; };
+
+    bool compare(const MPI_Comm& other) const;
+
+    bool operator==(const communicator& rhs) { return compare(rhs.get()); }
+
+    ~communicator();
+
+private:
+    MPI_Comm comm_;
+    int size_{};
+    int rank_{};
+    int local_rank_{};
+};
+
+
+/**
+ * A type helper which can be used to create MPI_Datatype from other types.
+ */
+class mpi_type {
+public:
+    mpi_type(const int count, MPI_Datatype& old);
+    ~mpi_type();
+    const MPI_Datatype& get() const { return this->type_; }
+
+private:
+    MPI_Datatype type_{};
+};
+
+
+/**
+ * This class wraps the MPI_Window class with RAII functionality. Different
+ * create and lock type methods are setup with enums.
+ *
+ * MPI_Window is primarily used for one sided communication and this class
+ * provides functionalities to fence, lock, unlock and flush the communication
+ * buffers.
+ */
+template <typename ValueType>
+class window {
+public:
+    enum class win_type { allocate = 1, create = 2, dynamic_create = 3 };
+    enum class lock_type { shared = 1, exclusive = 2 };
+
+    window() : window_(MPI_WIN_NULL) {}
+    window(window& other) = default;
+    window& operator=(const window& other) = default;
+    window(window&& other) = default;
+    window& operator=(window&& other) = default;
+
+    window(ValueType* base, unsigned int size,
+           std::shared_ptr<const communicator> comm,
+           const int disp_unit = sizeof(ValueType),
+           info input_info = info(MPI_INFO_NULL),
+           win_type create_type = win_type::create);
+
+    MPI_Win get() { return this->window_; }
+
+    void fence(int assert = 0);
+
+    void lock(int rank, int assert = 0, lock_type lock_t = lock_type::shared);
+
+    void unlock(int rank);
+
+    void lock_all(int assert = 0);
+
+    void unlock_all();
+
+    void flush(int rank);
+
+    void flush_local(int rank);
+
+    void flush_all();
+
+    void flush_all_local();
+
+    ~window();
+
+private:
+    MPI_Win window_;
+};
+
+
+/**
+ * This function is used to synchronize between the ranks of a given
+ * communicator.
```

Comment:
```suggestion
 * communicator. It will issue an MPI_Barrier call.
```
Just to clarify that this is a barrier, which I think is the more common term for this in the MPI community.